### PR TITLE
Rework the Loadout Synchronizer

### PR DESCRIPTION
**Pull request by halgari**
Implementation of the structure and tests of #626. This does not convert exsting games over to the new syncronizer, that will be done in a future pull request. All of the code is in the ALoadoutSyncronizer, start there or in ALoadoutSyncronizerTests to see usage.

### CODE REVIEW

File: src/NexusMods.DataModel/Abstractions/EntityDictionary.cs
Comment (erri120) [inline src/NexusMods.DataModel/Abstractions/EntityDictionary.cs:1]: `EntityDictionary<TK, TV>` can be made `readonly`.

File: src/NexusMods.DataModel/Abstractions/EntityDictionary.cs [hunk 1]
```diff
@@ -238,6 +238,35 @@ public TState Walk<TState>(Func<TState, Entity, TState> visitor, TState initial)
...
+        {
+            // We've already processed colliding keys, so skip them.
+            if (_coll.ContainsKey(key)) continue;
+            var newValue = mergeFn(null, _store.Get<TV>(value));
+            if (newValue is not null)
+                result = result.With(key, newValue);
+        }
+
+        return result;
+    }
```
Comment (erri120) [inline src/NexusMods.DataModel/Abstractions/EntityDictionary.cs:268]: `result = result.With(key, newValue)` uses `ImmutableDictionary.SetItem`. Doing this inside two for-each loops and creating new immutable dictionaries on possible every iteration is not very efficient. `ImmutableDictionary.SetItems` can be used instead.

File: src/NexusMods.DataModel/Loadouts/DiskStateRegistry.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    /// <returns></returns>
+    public DiskState? GetState(LoadoutId loadoutId)
+    {
+        var iid = loadoutId.ToEntityId(EntityCategory.DiskState);
+        var data = _dataStore.GetRaw(iid);
+        if (data == null) return null;
+        using var ms = new MemoryStream(data);
+        using var compressed = new GZipStream(ms, CompressionMode.Decompress);
+        return JsonSerializer.Deserialize<DiskState>(compressed);
+    }
```
Comment (erri120) [inline src/NexusMods.DataModel/Loadouts/DiskStateRegistry.cs:59]: Might be worth creating a helper method for doing JSON+GZip if we're going to be using it more.

File: src/NexusMods.DataModel/LoadoutSynchronizer/DiskModificationPlan.cs
Comment (erri120) [inline src/NexusMods.DataModel/LoadoutSynchronizer/DiskModificationPlan.cs:1]: This entire class is unused, is this a left-over?

File: src/NexusMods.DataModel/LoadoutSynchronizer/ALoadoutSynchronizer.cs [hunk 3]
```diff
@@ -0,0 +1,561 @@
...
+            await foreach (var entry in _hashCache.IndexFolderAsync(location))
+            {
+                var gamePath = installation.LocationsRegister.ToGamePath(entry.Path);
+                if (prevState.TryGetValue(gamePath, out var prevEntry))
+                {
+                    // If the file has been modified outside of the app since the last apply, we need to ingest it.
+                    if (prevEntry.Value!.Hash != entry.Hash)
+                    {
+                        HandleNeedIngest(entry);
+                        throw new UnreachableException("HandleNeedIngest should have thrown");
```
Comment (erri120) [inline src/NexusMods.DataModel/LoadoutSynchronizer/ALoadoutSynchronizer.cs:132]: `HandleNeedIngest` is `virtual`, so implementations can not throw. Is that method supposed to only throw or do something else?

File: src/NexusMods.DataModel/LoadoutSynchronizer/ALoadoutSynchronizer.cs [hunk 4]
```diff
@@ -0,0 +1,561 @@
...
+
+        // Write the generated files (could be done in parallel)
+        foreach (var entry in toWrite)
+        {
+            await using var outputStream = entry.Key.Create();
+            var hash = await entry.Value.Write(outputStream);
+            if (hash == null)
+            {
+                outputStream.Position = 0;
+                hash = await outputStream.HashingCopyAsync(Stream.Null, CancellationToken.None, async x => {});
```
Comment (erri120) [inline src/NexusMods.DataModel/LoadoutSynchronizer/ALoadoutSynchronizer.cs:208]: `reportFn` can be `null`.

**Review by erri120 — CHANGES_REQUESTED** [review]

File: src/NexusMods.DataModel/LoadoutSynchronizer/ALoadoutSynchronizer.cs (hunk 3 above)
Comment (halgari) [reply src/NexusMods.DataModel/LoadoutSynchronizer/ALoadoutSynchronizer.cs:132]: Anything that overrides that method is supposed to throw, as at this point the state of the apply process is not recoverable after a ingest. I'll leave it for now, but we can easily rework that code in the future if it's a problem.

File: src/NexusMods.DataModel/Loadouts/DiskStateRegistry.cs (hunk 2 above)
Comment (halgari) [reply src/NexusMods.DataModel/Loadouts/DiskStateRegistry.cs:59]: I hope to remove that code sometime in the future when I do work on the datamodel, and at that point we'll probably look into some sort of better compression method as well.

**Review by erri120 — APPROVED** [review]

**Comment by Al12rs** [comment]
Can I have a look before merging?

File: src/NexusMods.DataModel/Loadouts/Mods/Mod.cs [hunk 5]
```diff
@@ -24,6 +24,21 @@ public record Mod : Entity, IHasEntityId<ModId>
...
+
+    /// <summary>
+    /// Category used for 'Save Games'.
+    /// </summary>
+    public const string SavesCategory = "Saved Games";
+
+    /// <summary>
+    /// Category used for 'Save Games'.
+    /// </summary>
+    public const string OverridesCategory = "Overrides";
```
Comment (Al12rs) [inline src/NexusMods.DataModel/Loadouts/Mods/Mod.cs:40]: Comment needs fixing

File: src/NexusMods.DataModel/LoadoutSynchronizer/FlattenedToLoadoutTransformer.cs [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+                        continue;
+                    }
+                    else
+                    {
+                        AddToValues(_moveFrom, prevPair.Value.Mod.Id, prevPair.Value.File);
+                        AddToValues(_moveTo, newPair.Mod.Id, newPair.File);
+                        continue;
+                    }
+
+                    continue;
```
Comment (Al12rs) [inline src/NexusMods.DataModel/LoadoutSynchronizer/FlattenedToLoadoutTransformer.cs:81]: This is Unreachable, can be removed

**Review by Al12rs — APPROVED** [review]

**Pull request merged**